azure/services/privatelinks/spec.go

```diff
@@ -0,0 +1,239 @@
+			ClusterName: s.ClusterName,
+			Lifecycle:   infrav1.ResourceLifecycleOwned,
+			Name:        pointer.String(s.Name),
+			Additional:  s.AdditionalTags,
+		})),
+	}
+
+	if len(s.AllowedSubscriptions) > 0 {
+		privateLinkToCreate.Visibility = &network.PrivateLinkServicePropertiesVisibility{
+			Subscriptions: &s.AllowedSubscriptions,
+		}
+	}
+	if len(s.AutoApprovedSubscriptions) > 0 {
+		privateLinkToCreate.AutoApproval = &network.PrivateLinkServicePropertiesAutoApproval{
+			Subscriptions: &s.AutoApprovedSubscriptions,
+		}
+	}
+
+	if existing != nil {
+		existingPrivateLink, ok := existing.(network.PrivateLinkService)
+		if !ok {
+			return nil, errors.Errorf("%T is not a network.PrivateLinkService", existing)
+		}
+
+		if isExistingUpToDate(existingPrivateLink, privateLinkToCreate) {
+			return nil, nil
+		}
+	}
+
+	return privateLinkToCreate, nil
+}
+
+func isExistingUpToDate(existing network.PrivateLinkService, wanted network.PrivateLinkService) bool {
+	// NAT IP configuration is not checked as it cannot be changed.
+
+	// Check load balancer configurations
+	var wantedFrontendIDs []string
+	for _, wantedFrontendIPConfig := range *wanted.LoadBalancerFrontendIPConfigurations {
+		wantedFrontendIDs = append(wantedFrontendIDs, *wantedFrontendIPConfig.ID)
+	}
+	var existingFrontendIDs []string
+	for _, existingFrontendIPConfig := range *existing.LoadBalancerFrontendIPConfigurations {
+		existingFrontendIDs = append(existingFrontendIDs, *existingFrontendIPConfig.ID)
+	}
+	if !equalStringSlicesIgnoreOrder(wantedFrontendIDs, existingFrontendIDs) {
+		return false
+	}
+
+	// Check proxy protocol config
+	if !pointer.BoolEqual(wanted.EnableProxyProtocol, existing.EnableProxyProtocol) {
+		return false
+	}
+
+	// Check allowed subscriptions
+	if !equalStructWithStringSlicesPtrIgnoreOrder(
+		wanted.Visibility,
+		existing.Visibility,
+		func(v *network.PrivateLinkServicePropertiesVisibility) *[]string {
+			return v.Subscriptions
+		}) {
+		return false
+	}
+
+	// Check auto-approved subscriptions
+	if !equalStructWithStringSlicesPtrIgnoreOrder(
+		wanted.AutoApproval,
+		existing.AutoApproval,
+		func(v *network.PrivateLinkServicePropertiesAutoApproval) *[]string {
+			return v.Subscriptions
+		}) {
+		return false
+	}
+
+	return true
+}
+
+func equalStructWithStringSlicesPtrIgnoreOrder[T any](a, b *T, getStringSlice func(*T) *[]string) bool {
+	if bothPointersAreNil(a, b) {
+		return true
+	}
+	if onlyOnePointerIsNil(a, b) {
+		return false
+	}
+
+	return equalStringSlicesPtrIgnoreOrder(getStringSlice(a), getStringSlice(b))
+}
+
+func equalStringSlicesPtrIgnoreOrder(a, b *[]string) bool {
+	if bothPointersAreNil(a, b) {
+		return true
+	}
+	if onlyOnePointerIsNil(a, b) {
+		return false
+	}
+
+	// both are different from nil
+	return equalStringSlicesIgnoreOrder(*a, *b)
+}
+
+func equalStringSlicesIgnoreOrder(a, b []string) bool {
+	slicesDiff := cmp.Diff(a, b, cmpopts.SortSlices(func(s1, s2 string) bool {
```

Comment:
WDYT about creating a set for each slice and doing the diff on that instead? That way, it would be in linear time.